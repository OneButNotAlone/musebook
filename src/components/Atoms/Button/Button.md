# Button

You click them, things happen... or the button is disabled and you don't click them.

## Usage

```jsx
    <Button>Default Button</Button>

    <Button secondary>Flat Secondary</Button>

    <Button disabled>Flat Disabled</Button>

    <Button secondary disabled>
        Flat Secondary Disabled
    </Button>

    <Button outlined>Flat Outlined</Button>
```

## TODO

[ ] Add Gradient Button