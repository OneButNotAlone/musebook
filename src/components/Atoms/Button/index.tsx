import * as React from 'react';

import { styled, as, Button as BaseButton, BlockProps } from 'reakit';

const BaseFlatButton = styled(BaseButton)`
    height: 36px;
    min-width: 64px;
    border-radius: 4px;
    background-color: #187EC2;
    text-transform: uppercase;
    padding: 0 16px;
    text-align: center;
    font-size: 18px;
    font-family: 'Bogle';
`;

const PrimaryFlatButton = styled(BaseFlatButton)`
    background-color: #187EC2;
    color: white;
    border: 1px solid #187EC2;

    &:hover {
        background-color: #0a4e8c;
        border: 1px solid #0a4e8c;
    }
`;

const SecondaryFlatButton = styled(BaseFlatButton)`
    background-color: white;
    border: 1px solid #666666;
    color: black;

    &:hover {
        border: 1px solid #187EC2;
        color: #187EC2;
    }
`;

const OutlinedFlatButton = styled(BaseFlatButton)`
    color: white;
    border: 1px solid white;

    &:hover {
        background-color: rgba(255, 255, 255, 0.5);
    }
`;

const DisabledFlatButton = styled(BaseFlatButton)`
    background-color: #ccc;
    color: #666666;
    border: 1px solid #ccc;
`;

const RoundedButton = styled(BaseButton)`
    background: linear-gradient(-134deg, #0CB0F3 0%, #116CAE 49%, #0A4E8C 100%);
    text-transform: uppercase;
    color: white;
    border-radius: 25.5px;
    min-width: 100px;
    max-width: 300px;
    height: 36px;
    border: none;

    &:hover {
        font-weight: bold;
    }
`;

const ComposedButton: React.SFC<
    {
        type?: 'flat' | 'outlined' | 'rounded';
        secondary?: boolean;
        disabled?: boolean;
    } & BlockProps
> = (props) => {
    const { type = 'flat', secondary, disabled, children } = props;

    switch (type) {
        case 'flat':
            if (disabled === true) {
                return <DisabledFlatButton {...props}>{children}</DisabledFlatButton>;
            }

            if (secondary === true) {
                return <SecondaryFlatButton {...props}>{children}</SecondaryFlatButton>;
            }

            return <PrimaryFlatButton {...props}>{children}</PrimaryFlatButton>;

        case 'outlined':
            return <OutlinedFlatButton {...props}>{children}</OutlinedFlatButton>;

        case 'rounded':
            return <RoundedButton {...props}>{children}</RoundedButton>;

        default:
            return <PrimaryFlatButton {...props}>{children}</PrimaryFlatButton>;
    }
};

export const Button = as('button')(ComposedButton);
export default Button;
