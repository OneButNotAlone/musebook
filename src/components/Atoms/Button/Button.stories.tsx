import * as React from 'react';
import { storiesOf } from '@storybook/react';

import { Button } from './';

import * as ButtonMarkdown from './Button.md';

storiesOf('Buttons', module)
    .add('Default', () => (
        <Button>Default Button</Button>
    ), { notes: ButtonMarkdown })
    .add('Secondary', () => (
        <Button secondary>Flat Secondary</Button>
    ), { notes: ButtonMarkdown })
    .add('Disabled', () => (
        <Button disabled>Flat Disabled</Button>
    ), { notes: ButtonMarkdown })
    .add('Flat Secondary', () => (
        <Button secondary disabled>
            Flat Secondary Disabled
        </Button>
    ), { notes: ButtonMarkdown })
    .add('Outlined', () => (
        <Button type="outlined">Flat Outlined</Button>
    ), { notes: ButtonMarkdown });
