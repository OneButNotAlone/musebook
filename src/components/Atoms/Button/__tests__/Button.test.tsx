import * as React from 'react';
import 'jest-styled-components';
import Button from '../index';
import * as renderer from 'react-test-renderer'


test('Button should render with correct default styles', () => {
    const tree = renderer.create(
        <Button />
    ).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('height', '36px')
    expect(tree).toHaveStyleRule('min-width', '64px')
    expect(tree).toHaveStyleRule('border', '1px solid #187EC2')
    expect(tree).toHaveStyleRule('border-radius', '4px')
    expect(tree).toHaveStyleRule('background-color', '#187EC2')
    expect(tree).toHaveStyleRule('color', 'white')
    expect(tree).toHaveStyleRule('text-transform', 'uppercase')
    expect(tree).toHaveStyleRule('padding', '0 16px')
    expect(tree).toHaveStyleRule('text-align', 'center')
    expect(tree).toHaveStyleRule('font-size', '18px')
    expect(tree).toHaveStyleRule('font-family', expect.stringContaining('Bogle'))
    expect(tree).toHaveStyleRule('background-color', '#0a4e8c', { modifier: ':hover' })
    expect(tree).toHaveStyleRule('border', '1px solid #0a4e8c', { modifier: ':hover' })
});

test('Button should render with correct rounded styles', () => {
    const tree = renderer.create(
        <Button type='rounded'/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('text-transform', 'uppercase')
    expect(tree).toHaveStyleRule('color', 'white')
    expect(tree).toHaveStyleRule('border-radius', '25.5px')
    expect(tree).toHaveStyleRule('min-width', '100px')
    expect(tree).toHaveStyleRule('max-width', '300px')
    expect(tree).toHaveStyleRule('height', '36px')
    expect(tree).toHaveStyleRule('border', 'none')
    expect(tree).toHaveStyleRule('font-weight', 'bold', { modifier: ':hover' })
});

test('Button should render with correct outlined styles', () => {
    const tree = renderer.create(
        <Button type='outlined'/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('color', 'white')
    expect(tree).toHaveStyleRule('border', '1px solid white')
});

test('Button should render with correct secondary styles', () => {
    const tree = renderer.create(
        <Button secondary />
    ).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('background-color', 'white')
    expect(tree).toHaveStyleRule('color', 'black')
    expect(tree).toHaveStyleRule('border', '1px solid #666666')
    expect(tree).toHaveStyleRule('color', '#187EC2', { modifier: ':hover' })
    expect(tree).toHaveStyleRule('border', '1px solid #187EC2', { modifier: ':hover' })
});

test('Button should render with correct disabled styles', () => {
    const tree = renderer.create(
        <Button disabled />
    ).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('background-color', '#ccc')
    expect(tree).toHaveStyleRule('color', '#666666')
    expect(tree).toHaveStyleRule('border', '1px solid #ccc')
});