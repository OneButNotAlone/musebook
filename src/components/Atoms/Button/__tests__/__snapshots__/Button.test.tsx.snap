// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render with correct default styles 1`] = `
.c2 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1 {
  height: 36px;
  min-width: 64px;
  border-radius: 4px;
  background-color: #187EC2;
  text-transform: uppercase;
  padding: 0 16px;
  text-align: center;
  font-size: 18px;
  font-family: 'Bogle';
}

.c0 {
  background-color: #187EC2;
  color: white;
  border: 1px solid #187EC2;
}

.c0:hover {
  background-color: #0a4e8c;
  border: 1px solid #0a4e8c;
}

<button
  className="c0 c1 c2"
/>
`;

exports[`Button should render with correct disabled styles 1`] = `
.c2 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1 {
  height: 36px;
  min-width: 64px;
  border-radius: 4px;
  background-color: #187EC2;
  text-transform: uppercase;
  padding: 0 16px;
  text-align: center;
  font-size: 18px;
  font-family: 'Bogle';
}

.c0 {
  background-color: #ccc;
  color: #666666;
  border: 1px solid #ccc;
}

<button
  className="c0 c1 c2"
  disabled={true}
/>
`;

exports[`Button should render with correct outlined styles 1`] = `
.c2 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1 {
  height: 36px;
  min-width: 64px;
  border-radius: 4px;
  background-color: #187EC2;
  text-transform: uppercase;
  padding: 0 16px;
  text-align: center;
  font-size: 18px;
  font-family: 'Bogle';
}

.c0 {
  color: white;
  border: 1px solid white;
}

.c0:hover {
  background-color: rgba(255,255,255,0.5);
}

<button
  className="c0 c1 c2"
  type="outlined"
/>
`;

exports[`Button should render with correct rounded styles 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c0 {
  background: linear-gradient(-134deg,#0CB0F3 0%,#116CAE 49%,#0A4E8C 100%);
  text-transform: uppercase;
  color: white;
  border-radius: 25.5px;
  min-width: 100px;
  max-width: 300px;
  height: 36px;
  border: none;
}

.c0:hover {
  font-weight: bold;
}

<button
  className="c0 c1"
  type="rounded"
/>
`;

exports[`Button should render with correct secondary styles 1`] = `
.c2 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1 {
  height: 36px;
  min-width: 64px;
  border-radius: 4px;
  background-color: #187EC2;
  text-transform: uppercase;
  padding: 0 16px;
  text-align: center;
  font-size: 18px;
  font-family: 'Bogle';
}

.c0 {
  background-color: white;
  border: 1px solid #666666;
  color: black;
}

.c0:hover {
  border: 1px solid #187EC2;
  color: #187EC2;
}

<button
  className="c0 c1 c2"
/>
`;
